<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Async Functions</title>
</head>

<body>

</body>

</html>





<script>
    // import '../stylesheets/main.scss';


    // async function hello() {
    // }


    const sing = async () => {
        throw "OH NO, PROBLEM! ahhhhhhhhhhhhhjhhjhjhjhjhhjhjh"
        return 'LA LA LA LA'
    }

    sing()
        .then(data => {
            console.log("PROMISE RESOLVED WITH:", data)
        })
        .catch(err => {
            console.log("OH NO, PROMISE REJECTED!")
            console.log(err)
        })

    // --------------------------------------// --------------------------------------


    const login = async (username, password) => {
        if (!username || !password) throw 'Missing Credentials' // どっちかがないなら、エラー
        if (password === 'corgifeetarecute') return 'WELCOME!' //　このパスで来たら成功
        throw 'Invalid Password' // otherwise gives error 
    }

    login('todd', 'corgifeetarecute') // this matches above, so it gives success!!
        .then(msg => {
            console.log("LOGGED IN!")
            console.log(msg)
        })
        .catch(err => {
            console.log("ERROR!")
            console.log(err)
        })



    // --------------------------------------// --------------------------------------


    const delayedColorChange = (color, delay) => {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                document.body.style.backgroundColor = color;
                resolve();
            }, delay)
        })
    }

    // delayedColorChange('red', 1000)
    //     .then(() => delayedColorChange('orange', 1000))
    //     .then(() => delayedColorChange('yellow', 1000))
    //     .then(() => delayedColorChange('green', 1000))
    //     .then(() => delayedColorChange('blue', 1000))
    //     .then(() => delayedColorChange('indigo', 1000))
    //     .then(() => delayedColorChange('violet', 1000))

    //同じ秒数にしているが、 awaitのおかげで、待ってたから次のものが実行される
    async function rainbow() {
        await delayedColorChange('red', 1000), console.log("HELOOOOO")
        await delayedColorChange('orange', 1000)
        await delayedColorChange('yellow', 1000)
        await delayedColorChange('green', 1000)
        await delayedColorChange('blue', 1000)
        await delayedColorChange('indigo', 1000)
        await delayedColorChange('violet', 1000)
        return "ALL DONE!" // ここでpromise is resolved !!!
    }

    // rainbow().then(() => console.log("END OF RAINBOW!"))

    // 上で定義してるものが全部終わってから実行される　（This is another async　function )
    async function printRainbow() {
        await rainbow();
        console.log("END OF RAINBOW!")
    }

    printRainbow();






    // --------------------------------------// --------------------------------------

    const fakeRequest = (url) => {
        return new Promise((resolve, reject) => {
            const delay = Math.floor(Math.random() * (4500)) + 500;
            setTimeout(() => {
                if (delay > 2000) {
                    reject('Connection Timeout :(   ohhhh  noooooo')
                } else {
                    resolve(`Here is your fake data from ${url}`)
                }
            }, delay)
        })
    }

    // 上で定義してるものが全部終わってから実行される　（This is another async　function )
    async function makeTwoRequests() {
        try {
            let data1 = await fakeRequest('/page1');
            console.log(data1);
            let data2 = await fakeRequest('/page2');
            console.log(data2);
        } catch (e) {
            console.log("CAUGHT AN ERROR!")
            console.log("error is:", e)
        }

    }

</script>